comm/comm.1
.Dd December 2008
.Dt COMM 1

.Sh NAME

.Nm comm
.Nd select or reject lines common to two files

.Sh SYNOPSIS

.Nm comm
.Ar [-123] file1 file2

.Sh DESCRIPTION

    The comm utility shall read file1 and file2, which should be ordered in
the current collating sequence, and produce three text columns as output:
lines only in file1, lines only in file2, and lines in both files.

    If the lines in both files are not ordered according to the collating
sequence of the current locale, the results are unspecified.

.Sh OPTIONS

    The comm utility shall conform to the Base Definitions volume of IEEE
Std 1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -1
        Suppress the output column of lines unique to file1.
    -2
        Suppress the output column of lines unique to file2.
    -3
        Suppress the output column of lines duplicated in file1 and file2.

.Sh OPERANDS

    The following operands shall be supported:

    file1
        A pathname of the first file to be compared. If file1 is '-', the
standard input shall be used.
    file2
        A pathname of the second file to be compared. If file2 is '-', the
standard input shall be used.

    If both file1 and file2 refer to standard input or to the same FIFO
special, block special, or character special file, the results are
undefined.

.Sh STDIN

    The standard input shall be used only if one of the file1 or file2
operands refers to standard input. See the INPUT FILES section.

.Sh INPUT FILES

    The input files shall be text files.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of comm:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_COLLATE
        Determine the locale for the collating sequence comm expects to have
been used when the input files were sorted.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments and input files).
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    The comm utility shall produce output depending on the options selected.
If the -1, -2, and -3 options are all selected, comm shall write nothing to
standard output.

    If the -1 option is not selected, lines contained only in file1 shall be
written using the format:

    "%s\n", <line in file1>

    If the -2 option is not selected, lines contained only in file2 are
written using the format:

    "%s%s\n", <lead>, <line in file2>

    where the string <lead> is as follows:

    <tab>
        The -1 option is not selected.
    null string
        The -1 option is selected.

    If the -3 option is not selected, lines contained in both files shall be
written using the format:

    "%s%s\n", <lead>, <line in both>

    where the string <lead> is as follows:

    <tab><tab>
        Neither the -1 nor the -2 option is selected.
    <tab>
        Exactly one of the -1 and -2 options is selected.
    null string
        Both the -1 and -2 options are selected.

    If the input files were ordered according to the collating sequence of
the current locale, the lines written shall be in the collating sequence of
the original lines.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        All input files were successfully output as specified.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    If the input files are not properly presorted, the output of comm might
not be useful.

.Sh EXAMPLES

    If a file named xcu contains a sorted list of the utilities in this
volume of IEEE Std 1003.1-2001, a file named xpg3 contains a sorted list of
the utilities specified in the X/Open Portability Guide, Issue 3, and a file
named svid89 contains a sorted list of the utilities in the System V
Interface Definition Third Edition:

    comm -23 xcu xpg3 | comm -23 - svid89

    would print a list of utilities in this volume of IEEE Std 1003.1-2001
not specified by either of the other documents:

    comm -12 xcu xpg3 | comm -12 - svid89

    would print a list of utilities specified by all three documents, and:

    comm -12 xpg3 svid89 | comm -23 - xcu

    would print a list of utilities specified by both XPG3 and the SVID, but
not specified in this volume of IEEE Std 1003.1-2001.

.Sh RATIONALE

    None.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    cmp, diff, sort, uniq



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



comm/comm.cc
/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#ifndef _GNU_SOURCE
#define _GNU_SOURCE		/* for setlinebuf(3), fgets_unlocked(3) */
#endif

#include <sys/types.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <argp.h>
#include <libpu.h>


static const char doc[] =
N_("comm - compare two sorted files line by line");

static const char args_doc[] = N_("file1 file2");

static struct argp_option options[] = {
	{ NULL, '1', NULL, 0,
	  N_("Suppress the output column of lines unique to file1.") },
	{ NULL, '2', NULL, 0,
	  N_("Suppress the output column of lines unique to file2.") },
	{ NULL, '3', NULL, 0,
	  N_("Suppress the output column of lines duplicated in file1 and file2.") },
	{ }
};

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, args_doc, doc };

enum parse_options_bits {
	OPT_FILE1			= (1 << 0),
	OPT_FILE2			= (1 << 1),
	OPT_DUP				= (1 << 2),
};

static int outmask;
static const char *file1, *file2, *lead_dup, *lead_f2;
static char buf1[LINE_MAX + 1], buf2[LINE_MAX + 1];

static void line_out(int ltype, const char *line)
{
	int rc;

	if (ltype & outmask)
		return;

	switch (ltype) {
	case OPT_FILE1:
		rc = fputs(line, stdout);
		break;
	case OPT_FILE2:
		rc = fprintf(stdout, "%s%s", lead_f2, line);
		break;
	case OPT_DUP: {
		rc = fprintf(stdout, "%s%s", lead_dup, line);
		break;
	}
	default:
		rc = 0;
		break;
	}

	if (rc < 0)
		perror("stdout");
}

static int compare_files(void)
{
	FILE *f1, *f2;
	int want1, want2, rc;
	char *l1, *l2;

	rc = ro_file_open(&f1, file1);
	if (rc)
		return 1;

	rc = ro_file_open(&f2, file2);
	if (rc)
		goto out;

	setlinebuf(f1);
	setlinebuf(f2);

	l1 = l2 = NULL;
	want1 = want2 = 1;
	while (want1 || want2) {
		if (want1 && (!l1)) {
			l1 = fgets_unlocked(buf1, sizeof(buf1), f1);
			if (!l1) {
				if (ferror(f1)) {
					perror(file1);
					rc = 1;
					goto out2;
				}
				want1 = 0;
			}
		}
		if (want2 && (!l2)) {
			l2 = fgets_unlocked(buf2, sizeof(buf2), f2);
			if (!l2) {
				if (ferror(f2)) {
					perror(file2);
					rc = 1;
					goto out2;
				}
				want2 = 0;
			}
		}

		if (!l1 && !l2)
			break;

		if (!l1) {
			line_out(OPT_FILE2, l2);
			l2 = NULL;
		} else if (!l2) {
			line_out(OPT_FILE1, l1);
			l1 = NULL;
		} else {
			int cmp = strcoll(l1, l2);
			if (cmp < 0) {		/* l1 < l2 */
				line_out(OPT_FILE1, l1);
				l1 = NULL;
			}
			else if (cmp > 0) {	/* l1 > l2 */
				line_out(OPT_FILE2, l2);
				l2 = NULL;
			}
			else {			/* l1 == l2 */
				line_out(OPT_DUP, l1);
				l1 = l2 = NULL;
			}
		}
	}

out2:
	if (fclose(f2))
		perror(file2);
out:
	if (fclose(f1))
		perror(file1);
	return rc;
}

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	switch (key) {
	case '1':
		outmask |= OPT_FILE1;
		break;
	case '2':
		outmask |= OPT_FILE2;
		break;
	case '3':
		outmask |= OPT_DUP;
		break;

	case ARGP_KEY_ARG:
		switch(state->arg_num) {
		case 0:		file1 = arg; break;
		case 1:		file2 = arg; break;
		default:	argp_usage (state); break; /* too many args */
		}
		break;

	case ARGP_KEY_END:
		if (state->arg_num < 2)		/* not enough args */
			argp_usage (state);
		break;

	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

int main (int argc, char *argv[])
{
	pu_init();

	error_t argp_rc = argp_parse(&argp, argc, argv, 0, NULL, NULL);
	if (argp_rc) {
		fprintf(stderr, _("%s: argp_parse failed: %s\n"),
			argv[0], strerror(argp_rc));
		return EXIT_FAILURE;
	}

	lead_f2 = (outmask & OPT_FILE1) ? "" : "\t";
	if ((outmask & (OPT_FILE1 | OPT_FILE2)) == 0)
		lead_dup = "\t\t";
	else if ((outmask & (OPT_FILE1 | OPT_FILE2)) == (OPT_FILE1 | OPT_FILE2))
		lead_dup = "";
	else
		lead_dup = "\t";

	return compare_files();
}

